# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1425348013 18000
#      Mon Mar 02 21:00:13 2015 -0500
# Node ID e1d7bd38b82b0d28320759aa885ff1124975f924
# Parent  966a1933f3fa0489b7184674859c53ba26ef3316
make pan, rotate3d, and zoom properly toggle state

* pan.m (update_mouse_mode): New subfunction.  Also set __mouse_mode__
property when toggling state.
* rotate3d.m: Likewise.
* zoom.m: Likewise.

diff --git a/scripts/plot/util/pan.m b/scripts/plot/util/pan.m
--- a/scripts/plot/util/pan.m
+++ b/scripts/plot/util/pan.m
@@ -71,16 +71,17 @@ function pan (varargin)
   if (nargs == 0)
     pm = get (hfig, "__pan_mode__");
     if (strcmp (pm.Enable, "on"))
       pm.Enable = "off";
     else
       pm.Enable = "on";
     endif
     set (hfig, "__pan_mode__", pm);
+    update_mouse_mode (hfig, pm.Enable);
   elseif (nargs == 1)
     arg = varargin{1};
     if (ischar (arg))
       switch (arg)
         case {"on", "off", "xon", "yon"}
           pm = get (hfig, "__pan_mode__");
           switch (arg)
             case {"on", "off"}
@@ -89,29 +90,31 @@ function pan (varargin)
             case "xon"
               pm.Enable = "on";
               pm.Motion = "horizontal";
             case "yon"
               pm.Enable = "on";
               pm.Motion = "vertical";
           endswitch
           set (hfig, "__pan_mode__", pm);
-          if (strcmp (arg, "off"))
-            set (hfig, "__mouse_mode__", "none");
-          else
-            ## FIXME: Is there a better way other than calling these
-            ## functions to set the other mouse mode Enable fields to
-            ## "off"?
-            rotate3d ("off");
-            zoom ("off");
-            set (hfig, "__mouse_mode__", "pan");
-          endif
-
+          update_mouse_mode (hfig, arg);
         otherwise
           error ("pan: unrecognized option '%s'", arg);
       endswitch
     else
       error ("pan: wrong type argument '%s'", class (arg));
     endif
   endif
 
 endfunction
 
+function update_mouse_mode (hfig, arg)
+  if (strcmp (arg, "off"))
+    set (hfig, "__mouse_mode__", "none");
+  else
+    ## FIXME: Is there a better way other than calling these
+    ## functions to set the other mouse mode Enable fields to
+    ## "off"?
+    rotate3d ("off");
+    zoom ("off");
+    set (hfig, "__mouse_mode__", "pan");
+  endif
+endfunction
diff --git a/scripts/plot/util/rotate3d.m b/scripts/plot/util/rotate3d.m
--- a/scripts/plot/util/rotate3d.m
+++ b/scripts/plot/util/rotate3d.m
@@ -66,41 +66,44 @@ function rotate3d (varargin)
   if (nargs == 0)
     rm = get (hfig, "__rotate_mode__");
     if (strcmp (rm.Enable, "on"))
       rm.Enable = "off";
     else
       rm.Enable = "on";
     endif
     set (hfig, "__rotate_mode__", rm);
+    update_mouse_mode (hfig, rm.Enable);
   elseif (nargs == 1)
     arg = varargin{1};
     if (ischar (arg))
       switch (arg)
         case {"on", "off"}
           rm = get (hfig, "__rotate_mode__");
           switch (arg)
             case {"on", "off"}
               rm.Enable = arg;
               rm.Motion = "both";
           endswitch
           set (hfig, "__rotate_mode__", rm);
-          if (strcmp (arg, "off"))
-            set (hfig, "__mouse_mode__", "none");
-          else
-            ## FIXME: Is there a better way other than calling these
-            ## functions to set the other mouse mode Enable fields to
-            ## "off"?
-            pan ("off");
-            zoom ("off");
-            set (hfig, "__mouse_mode__", "rotate");
-          endif
-
+          update_mouse_mode (hfig, arg);
         otherwise
           error ("rotate3d: unrecognized option '%s'", arg);
       endswitch
     else
       error ("rotate3d: wrong type argument '%s'", class (arg));
     endif
   endif
 
 endfunction
 
+function update_mouse_mode (hfig, arg)
+  if (strcmp (arg, "off"))
+    set (hfig, "__mouse_mode__", "none");
+  else
+    ## FIXME: Is there a better way other than calling these
+    ## functions to set the other mouse mode Enable fields to
+    ## "off"?
+    pan ("off");
+    zoom ("off");
+    set (hfig, "__mouse_mode__", "rotate");
+  endif
+endfunction
diff --git a/scripts/plot/util/zoom.m b/scripts/plot/util/zoom.m
--- a/scripts/plot/util/zoom.m
+++ b/scripts/plot/util/zoom.m
@@ -87,16 +87,17 @@ function zoom (varargin)
   if (nargs == 0)
     zm = get (hfig, "__zoom_mode__");
     if (strcmp (zm.Enable, "on"))
       zm.Enable = "off";
     else
       zm.Enable = "on";
     endif
     set (hfig, "__zoom_mode__", zm);
+    update_mouse_mode (hfig, zm.Enable);
   elseif (nargs == 1)
     arg = varargin{1};
     if (isnumeric (arg))
       factor = arg;
       switch (numel (factor))
         case 2
           xfactor = factor(1);
           yfactor = factor(2);
@@ -136,49 +137,49 @@ function zoom (varargin)
             case "xon"
               zm.Enable = "on";
               zm.Motion = "horizontal";
             case "yon"
               zm.Enable = "on";
               zm.Motion = "vertical";
           endswitch
           set (hfig, "__zoom_mode__", zm);
-          if (strcmp (arg, "off"))
-            set (hfig, "__mouse_mode__", "none");
-          else
-            ## FIXME: Is there a better way other than calling these
-            ## functions to set the other mouse mode Enable fields to
-            ## "off"?
-            pan ("off");
-            rotate3d ("off");
-            set (hfig, "__mouse_mode__", "zoom");
-          endif
-
+          update_mouse_mode (hfig, arg);
         case "out"
           cax = get (hfig, "currentaxes");
           if (! isempty (cax))
             __zoom__ (cax, "out");
           endif
-
         case "reset"
           cax = get (hfig, "currentaxes");
           if (! isempty (cax))
             __zoom__ (cax, "reset");
           endif
-
         otherwise
           error ("zoom: unrecognized option '%s'", arg);
       endswitch
     else
       error ("zoom: wrong type argument '%s'", class (arg));
     endif
   endif
 
 endfunction
 
+function update_mouse_mode (hfig, arg)
+  if (strcmp (arg, "off"))
+    set (hfig, "__mouse_mode__", "none");
+  else
+    ## FIXME: Is there a better way other than calling these
+    ## functions to set the other mouse mode Enable fields to
+    ## "off"?
+    pan ("off");
+    rotate3d ("off");
+    set (hfig, "__mouse_mode__", "zoom");
+  endif
+endfunction
 
 %!demo
 %! clf;
 %! sombrero ();
 %! pause (1);
 %! %% zoom in by a factor of 2
 %! zoom (2);
 %! pause (1);
